### DISCUSSION TITLE
Fine grained auth controls

### DISCUSSION THREAD

**Discussion by mosnicholas** [supabase/supabase#15413] [Questions]
I’ve set up my supabase instance to be sign up only. This is because we need to set up a new subdomain for each business. However, once we’ve got a subdomain set up, I’d like to let anyone in the company (ie with a certain email suffix) be able to sign in. 

Is there a way to keep it be sign up only, but allow anyone who matches certain criteria to go through? Or is that something that we’d have to do in the application layer?

**Comment by GaryAustin1** [comment]
If you are using the user management starter auth.users trigger with function approach you can just check for email in that function and raise an
 exception if it does not meet your test.  If not using the starter, you can still use the same approach with the function and do the test without a profile table.

**Reply by mosnicholas** [reply]
oh interesting -- could you point me to an example if any come to mind @GaryAustin1 ? :) I'm not sure I entirely follow how it would be set up, but it sounds promising.

I'm not sure what the `user management starter` is, but I haven't changed any auth settings, so I presume I should be using it?

**Reply by GaryAustin1** [reply]
user management starter is the guide for setting up a public profile table.
The code is in the examples in SQL editor...
![image](https://github.com/supabase/supabase/assets/54564956/ded3f43a-574b-4843-8689-60965054ca0b)
That shows a function you could add the email check in, or modify if you are not using the profile table.
I don't have a code snippet handy, but I suspect there is an example of this out there.
Also one of the AI's might provide a good example as it is all Postgres.

**Reply by mosnicholas** [reply]
Got it. I built that out a while ago to create Profiles, wish I'd known that existed :D 

In this model, users can attempt to sign up, but if they aren't "approved" the sign up will fail because the Profile creation step will fail?

It feels like it will work, but I'm worried it might be a little fickle (ie. if for whatever reason we should move away from using the profiles, it'll stop gating access) :/

is there any other callback that the auth function makes on sign up that we could run checks on? Would be super valuable to support that generally :D

**Reply by GaryAustin1** [reply]
There is no callback from auth.  You only have table operations to stick triggers on.   You can use an auth.users trigger without profiles.  You just do the insert trigger and check the email, if it does not match then return null intead of new (or raise an exception) and the user will fail to be created.  Of course you should do a filter in the client code for nice error message and just use this as a security backstop.
